File: LectureCodes/Filtering/circeig/Eigen/circeig.cpp
# include <iostream>
# include <Eigen/Dense>
# include <Eigen/Eigenvalues>
# include <figure/figure.hpp>
# include "circul.hpp"
using Eigen::VectorXd;
using Eigen::MatrixXd;
using Eigen::MatrixXcd; // complex for eigenvectors

// @brief Plot real and imaginary part of eigenvectors of n x n 
//        circulant matrix. 'ind' is just for the plot title
void circeig(const int n, const int ind) {
  const std::string title = "Circ matrix " + std::to_string(ind),
                    saveas = "circeig" + std::to_string(ind) + "ev";
  // get random circulant matrix and get eigenvectors
  MatrixXd C; circul(C, VectorXd::Random(n));
  Eigen::EigenSolver<MatrixXd> eig(C);
  MatrixXcd V = eig.eigenvectors();

  for (int j = 0; j < n; ++j) {
    MatrixXd data(n, 2); // prepare bar data
    // try: data << V.col(j).real(), V.col(j).imag();
    data.col(0) = V.col(j).real(); data.col(1) = V.col(j).imag();
    mgl::Figure fig;  fig.title(title);
    fig.bar(data, "br"); 
    fig.xlabel("vector component index");
    fig.ylabel("vector component value");
    fig.addlabel("real part", "b");
    fig.addlabel("imaginary part", "r");
    fig.legend(0,0);
    fig.save(saveas + std::to_string(j));
  }
}

int main() {
  circeig(8, 1);
  circeig(8, 2);
  return 0;
}

File: LectureCodes/Filtering/circeig/Matlab/circeig.m
function circeig
% Computing the eigenvectors of a \Magenta{random} circulant matrix
n = 8;
% The MATLAB way to generate a circulant matrix
C = gallery('circul',rand(n,1)); [V1,D1] = eig(C);

for j=1:n
  figure; bar(1:n,[real(V1(:,j)),imag(V1(:,j))],1,'grouped');
  title(sprintf('Circulant matrix 1, eigenvector %d',j));
  xlabel('{\bf vector component index}','fontsize',14);
  ylabel('{\bf vector component value}','fontsize',14);   
  legend('real part','imaginary part','location','southwest'); 
  print('-depsc2',sprintf('../PICTURES/circeig1ev%d.eps',j));
end

C = gallery('circul',rand(n,1)); [V2,D2] = eig(C);

for j=1:n
  figure; bar(1:n,[real(V2(:,j)),imag(V2(:,j))],1,'grouped');
  title(sprintf('Circulant matrix 2, eigenvector %d',j));
  xlabel('{\bf vector component index}','fontsize',14);
  ylabel('{\bf vector component value}','fontsize',14);   
  legend('real part','imaginary part','location','southwest'); 
  print('-depsc2',sprintf('../PICTURES/circeig2ev%d.eps',j));
end

figure; plot(1:n,real(diag(D1)),'r+',1:n,imag(diag(D1)),'b+',...
	     1:n,real(diag(D2)),'m*',1:n,imag(diag(D2)),'k*');	
ax = axis; axis([0 n+1 ax(3) ax(4)]);
xlabel('{\bf index of eigenvalue}','fontsize',14);
ylabel('{\bf eigenvalue}','fontsize',14);   
legend('C_1: real(ev)', 'C_1: imag(ev)', 'C_2: real(ev)', 'C_2: imag(ev)', 'location','northeast'); 

print -depsc2 '../PICTURES/circeigev.eps';

